### ISSUE TITLE
[BUG] `@basic_value` default constructor is not `implicit`

### ISSUE THREAD

**Issue by JohelEGP** [hsutter/cppfront#553]
<a id="title"/>

**[T](#title)itle**: `@basic_value` default constructor is not `implicit`.

<a id="desc"/>

**[D](#desc)escription**:

We expect a value to be assignable from an empty argument list, like `x = ()`.

<a id="rep"/>
<a id="cpp2"/>

**[M](#rep)inimal reproducer** (<https://cpp2.godbolt.org/z/ahM4Pcdbq>):

```Cpp2
t: @basic_value type = { }
u: type = {
  a: t = ();
  operator=: (out this, x: i32) = _ = x;
}
main: () = { }
```

<a id="cmd"/>
<details><summary>
<b><a href="#cmd">C</a>ommands</b>:
</summary>

```bash
cppfront main.cpp2
clang++17 -std=c++23 -stdlib=libc++ -lc++abi -pedantic-errors -Wall -Wextra -Wconversion -I . main.cpp
```
</details>

<a id="expected"/>

**[E](#expected)xpected result**: `public: t();`

<a id="actual"/>

**[A](#actual)ctual result and error**: `public: explicit t();`

<a id="cpp1"/>
<details><summary>
<a href="#cpp1">C</a>pp2 lowered to Cpp1:
</summary>

```C++


//=== Cpp2 type declarations ====================================================


#include "cpp2util.h"

class t;
class u;
  

//=== Cpp2 type definitions and function declarations ===========================

class t {
public: t(t const& that);

public: auto operator=(t const& that) -> t& ;
public: t(t&& that) noexcept;
public: auto operator=(t&& that) noexcept -> t& ;
public: explicit t();

};
class u {
  private: t a {}; 
  public: explicit u(cpp2::in<cpp2::i32> x);
  public: auto operator=(cpp2::in<cpp2::i32> x) -> u& ;

  public: u(u const&) = delete; /* No 'that' constructor, suppress copy */
  public: auto operator=(u const&) -> void = delete;
};
auto main() -> int;


//=== Cpp2 function definitions =================================================



t::t(t const& that){}
auto t::operator=(t const& that) -> t& {
                                return *this;}
t::t(t&& that) noexcept{}
auto t::operator=(t&& that) noexcept -> t& {
                                return *this;}
t::t(){}
  u::u(cpp2::in<cpp2::i32> x) { (void) x;  }
  auto u::operator=(cpp2::in<cpp2::i32> x) -> u&  { 
                                  a = {};
  (void) x;
                                  return *this;
   }

auto main() -> int{}
```
</details>

<a id="out"/>
<a href="#out">O</a>utput:

```output
main.cpp2:5:37: error: no viable overloaded '='
    5 |                                   a = {};
      |                                   ~ ^ ~~
```

**Comment by JohelEGP** [comment]
I noticed that this happens to work (<https://cpp2.godbolt.org/z/Wof3Gfsd7>):
```Cpp2
t: @basic_value type = {
  x: i32 = ();
  operator=: (out this, y: i32) = _ = y;
  // private operator=: (inout this, y: i32) = _ = y;
}
u: type = {
  a: t = ();
  operator=: (out this, x: i32) = _ = x;
}
main: () = { }
```
And it's because it actually invokes the "explicit" assignment (<https://github.com/hsutter/cppfront/issues/468#issuecomment-1627541323>).
If we uncomment the commented line to opt-out of it (<https://github.com/hsutter/cppfront/issues/468#issuecomment-1650115973>),
then Clang's error message makes it evident:
```output
main.cpp2:9:37: error: 'operator=' is a private member of 't'
    9 |                                   a = {};
      |                                   ~ ^ ~~
```
This is similar to #487's quote of <https://github.com/hsutter/cppfront/issues/451#issuecomment-1544560110>.

**Comment by JohelEGP** [comment]
See <https://cpp2.godbolt.org/z/8fE3sGxvn>:
- Such types still model `std::semiregular`.
- _The_ fix probably is to explicitly write `operator=: (inout this, _: i32) = a = _;`.
- Having the [NSDMI][] be `t()` also "fixes" it, but can introduce performance regressions (just like #451).
  * This easier "fix" reminds me of the discussion at #475 on the complexity of what's going on with the [SFM][]s.

[ABI]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#abi-api
[ADL]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#adl
[CNTTP]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#cnttp
[CPO]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#cpo
[CRTP]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#crtp
[CTAD]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#ctad
[WG]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#cwg-ewg-ewgi-lewg-lewgi-lwg
[SG]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#cwg-ewg-ewgi-lewg-lewgi-lwg
[ICE]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#ice
[IFNDR]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#ifndr
[II]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#iile
[NSDMI]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#nsdmi
[NTTP]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#nttp
[ODR]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#odr
[PIMPL]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#pimpl
[QoI]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#qoi
[RVO]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#rvo-nrvo-urvo
[SBO]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#sbo-soo-sso
[SCARY]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#scary-iterators
[SFINAE]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#sfinae
[SMF]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#smf
[STL]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#stl
[TMP]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#tmp
[TU]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#tu
[UB]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#ub
[UDL]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#udl
[UFCS]: https://quuxplusone.github.io/blog/2019/08/02/the-tough-guide-to-cpp-acronyms/#ufcs

**Comment by JohelEGP** [comment]
But aggregates of `basic_value`s with the generated default constructor
can't be value initialized or initialized from an empty list (<https://cpp2.godbolt.org/z/f76WbYv4P>):
```Cpp2
t: @basic_value type = { }
u: @struct type = {
  a: t;
}
main: () = {
  _ = :u = ();
}
```
```output
main.cpp2:6:23: error: chosen constructor is explicit in copy-initialization
    6 |   static_cast<void>(u{});
      |                       ^
main.cpp2:9:4: note: explicit constructor declared here
    9 | t::t(){}
      |    ^
main.cpp2:3:13: note: in implicit initialization of field 'a' with omitted initializer
    3 |   public: t a; 
      |             ^
1 error generated.
```
